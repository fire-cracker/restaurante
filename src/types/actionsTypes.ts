export interface Action<T> {
  type: ActionType;
  payload?: T;
}

export enum ActionType {
  SIGNUP_REQUEST_PENDING = 'SIGNUP_REQUEST_PENDING',
  SIGNUP_REQUEST_SUCCESS = 'SIGNUP_REQUEST_SUCCESS',
  SIGNUP_REQUEST_FAILED = 'SIGNUP_REQUEST_FAILED',
  
  LOGIN_REQUEST_PENDING = 'LOGIN_REQUEST_PENDING',
  LOGIN_REQUEST_SUCCESS = 'LOGIN_REQUEST_SUCCESS',
  LOGIN_REQUEST_FAILED = 'LOGIN_REQUEST_FAILED',

  GET_PROFILE_REQUEST_PENDING = 'GET_PROFILE_REQUEST_PENDING',
  GET_PROFILE_REQUEST_SUCCESS = 'GET_PROFILE_REQUEST_SUCCESS',
  GET_PROFILE_REQUEST_FAILED = 'GET_PROFILE_REQUEST_FAILED',

  LOGOUT_REQUEST_SUCCESS = 'LOGOUT_REQUEST_SUCCESS',

  SET_USER_STATE = 'SET_USER_STATE'
}