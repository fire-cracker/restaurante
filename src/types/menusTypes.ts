export interface IMenu {
  id: number
  name: string
  image: string
  price: number
  recipe: string
  type: string
}

export interface IMenuState {
  menus: IMenu | null
  count: number
  fetching: boolean
  fetched: boolean
}
