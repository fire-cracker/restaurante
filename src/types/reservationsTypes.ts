export interface IReservation {
  id: number
  userId: string
  date: string
  time: string
  type: string
  price: number
  persons: number
}
export interface INewReservation {
  date: string
  time: string
  type: string
  persons: number
}

export interface IReservationState {
  reservation: IReservation | null
  fetching: boolean
  fetched: boolean
}

export interface IStripeCharge {
  id: string
  object: string
  amount: 2000
  paid: boolean
  receipt_email: string
  receipt_number: string
  receipt_url: string
  status: string
}
