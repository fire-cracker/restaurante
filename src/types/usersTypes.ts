export interface IUser {
  id: string
  username: string
  email: string
  role: 'customer' | 'admin'
}

export interface IUserState {
  user: IUser | null
  logingIn: boolean
  isLoggedIn: boolean
}