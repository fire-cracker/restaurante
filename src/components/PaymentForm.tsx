import React, { FC, useState, ReactElement } from 'react'
import { CardElement, useStripe, useElements } from '@stripe/react-stripe-js'
import { Token, StripeError } from '@stripe/stripe-js'
import ClipLoader from 'react-spinners/ClipLoader'
import { INewReservation, IStripeCharge } from '../types/reservationsTypes'

const cardStyle = {
  style: {
    base: {
      color: '#32325d',
      fontFamily: 'Arial, sans-serif',
      fontSmoothing: 'antialiased',
      fontSize: '16px',
      '::placeholder': {
        color: '#aab7c4'
      }
    },
    invalid: {
      color: '#fa755a',
      iconColor: '#fa755a'
    }
  }
}

interface IProps {
  reservation: INewReservation | null
  addReservation: (
    reservation: INewReservation,
    stripeToken: string
  ) => Promise<IStripeCharge | undefined>
  history: any
}

const PaymentForm: FC<IProps> = ({ reservation, addReservation, history }): ReactElement => {
  const [error, setError] = useState('')
  const [processing, setProcessing] = useState(false)
  const [disabled, setDisabled] = useState(true)
  const [receiptUrl, setReceiptUrl] = useState('')
  const elements = useElements()
  const stripe = useStripe()

  const handleChange = async (event: any) => {
    setDisabled(event.empty)
    setError(event.error ? event.error.message : '')
  }

  const handleSubmit = async (event: any) => {
    event.preventDefault()
    if (!stripe || !elements) {
      return
    }
    setProcessing(true)
    const cardElement = elements.getElement(CardElement)
    const result = (await stripe.createToken(cardElement!)) as {
      token: Token
      error: StripeError
    }

    if (result.error) {
      setError(result.error.message!)
    } else {
      setError('')
      const stripeCharge = (await addReservation(reservation!, result.token.id)) as IStripeCharge
      if (stripeCharge) {
        setReceiptUrl(stripeCharge.receipt_url)
      } else {
        setError('Payment failed')
        history.push('/reservation')
      }
    }
    setProcessing(false)
  }

  if (receiptUrl) {
    return (
      <div className="success">
        <h2>Payment Successful!</h2>
        <a href={receiptUrl}>View Receipt </a>
      </div>
    )
  }

  return (
    <form className="payment-form" onSubmit={handleSubmit}>
      <CardElement id="card-element" options={cardStyle} onChange={handleChange} />
      <button disabled={processing || disabled} id="submit" className="bg-black text-darkkhaki">
        <span id="button-text">
          {processing ? (
            <ClipLoader size={30} color={'#00acc1'} loading={true} />
          ) : (
            `Pay $${reservation!.persons * 1000}`
          )}
        </span>
      </button>
      {error && (
        <div className="card-errors" role="alert">
          {error}
        </div>
      )}
    </form>
  )
}

export default PaymentForm
