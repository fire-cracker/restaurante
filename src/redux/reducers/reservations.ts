import { ActionType, Action } from '../../types/actionsTypes'
import { IReservationState } from '../../types/reservationsTypes'

const {
  ADD_RESERVATION_REQUEST_PENDING,
  ADD_RESERVATION_REQUEST_SUCCESS,
  ADD_RESERVATION_REQUEST_FAILED
} = ActionType

const initialState: IReservationState = {
  reservation: null,
  fetching: false,
  fetched: false
}
export default (state = initialState, action: Action<any>): IReservationState => {
  switch (action.type) {
    case ADD_RESERVATION_REQUEST_PENDING:
      return {
        ...state,
        fetching: true
      }

    case ADD_RESERVATION_REQUEST_SUCCESS:
      return {
        ...state,
        reservation: action.payload,
        fetching: false,
        fetched: true
      }

    case ADD_RESERVATION_REQUEST_FAILED:
      return {
        ...state,
        fetching: false,
        fetched: false
      }

    default:
      return state
  }
}
