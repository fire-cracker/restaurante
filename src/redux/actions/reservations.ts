import { Dispatch } from 'redux'
import { toast } from 'react-toastify'

import axios from '../../utils/axiosConfig'
import { ActionType, Action } from '../../types/actionsTypes'
import { IUser } from '../../types/usersTypes'
import { IMenu } from '../../types/menusTypes'

const {
  ADD_RESERVATION_REQUEST_PENDING,
  ADD_RESERVATION_REQUEST_SUCCESS,
  ADD_RESERVATION_REQUEST_FAILED
} = ActionType

const api = process.env.REACT_APP_API_URL

export const addReservationRequestPending = (): Action<any> => ({
  type: ADD_RESERVATION_REQUEST_PENDING
})

export const addReservationRequestSuccess = (data: IUser): Action<IUser> => ({
  type: ADD_RESERVATION_REQUEST_SUCCESS,
  payload: data
})

export const addReservationRequestFailed = (): Action<any> => ({
  type: ADD_RESERVATION_REQUEST_FAILED
})

export const addReservation = () => async (
  dispatch: Dispatch<Action<any>>
): Promise<{ menus: IMenu[]; count: number }> => {
  try {
    dispatch(addReservationRequestPending())
    const {
      data: { data }
    } = await axios.post(`${api}/reservations`)
    dispatch(addReservationRequestSuccess(data))
    return data
  } catch (error) {
    dispatch(addReservationRequestFailed())
    toast.error(error.response.data.data.message)
    throw error
  }
}
