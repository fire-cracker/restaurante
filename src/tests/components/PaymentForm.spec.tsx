import React from 'react'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Elements } from '@stripe/react-stripe-js'
import { loadStripe } from '@stripe/stripe-js'

import PaymentForm from '../../components/PaymentForm'
import { reservation } from '../mocks/reservations.mock'
import { INewReservation } from '../../types/reservationsTypes'
import * as mocks from '../mocks/stripe.mock'

const stripePromise = loadStripe('pk_test_s7dzKE4O2saVThp2USNgEFoW00hc0xxPft')

interface IProps {
  reservation: INewReservation
  history: any
  addReservation: (reservation: any, stripeToken: any) => any
}

describe('PaymentForm', () => {
  const props: IProps = {
    reservation: reservation,
    history: {},
    addReservation: jest.fn()
  }

  let mockStripe: any
  let mockElements: any
  let mockElement: any
  let simulateChange: any
  let simulateBlur: any
  let simulateFocus: any
  let simulateEscape: any
  let simulateReady: any
  let simulateClick: any

  beforeEach(() => {
    mockStripe = mocks.mockStripe()
    mockElements = mocks.mockElements()
    mockElement = mocks.mockElement()
    mockStripe.elements.mockReturnValue(mockElements)
    mockElements.create.mockReturnValue(mockElement)
    jest.spyOn(React, 'useLayoutEffect')
    mockElement.on = jest.fn((event, fn) => {
      switch (event) {
        case 'change':
          simulateChange = fn
          break
        case 'blur':
          simulateBlur = fn
          break
        case 'focus':
          simulateFocus = fn
          break
        case 'escape':
          simulateEscape = fn
          break
        case 'ready':
          simulateReady = fn
          break
        case 'click':
          simulateClick = fn
          break
        default:
          throw new Error('TestSetupError: Unexpected event registration.')
      }
    })
  })

  afterAll(() => {
    jest.clearAllMocks()
  })

  test('renders PaymentForm component', () => {
    const wrapper = render(
      <Elements stripe={null}>
        <PaymentForm {...props} />
      </Elements>
    )
    expect(wrapper).toMatchSnapshot()
  })

  test('should trigger onchange when card details is changed', async () => {
    const mockHandler = jest.fn()
    render(
      <Elements stripe={mockStripe}>
        {/* @ts-ignore */}
        <PaymentForm onChange={mockHandler} />
      </Elements>
    )
    const changeEventMock = Symbol('change')
    userEvent.type(mockElement, simulateChange(changeEventMock))
    expect(mockHandler).toHaveBeenCalledWith(changeEventMock)
  })

  test('should submit charge on click', async () => {
    const mockHandler = jest.fn()
    const { getByText } = render(
      <Elements stripe={mockStripe}>
        {/* @ts-ignore */}
        <PaymentForm onClick={mockHandler} />
      </Elements>
    )
    const payButton = getByText('Pay')
    userEvent.click(payButton)
    expect(mockHandler).toHaveBeenCalled()
  })
})
