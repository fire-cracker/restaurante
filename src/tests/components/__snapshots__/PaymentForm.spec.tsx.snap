// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentForm renders PaymentForm component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="p-0 jumbotron jumbotron-fluid"
      >
        <div
          class="container-fluid"
        >
          <div
            class="top-banner row"
          />
          <div
            class="reservation-wrapper px-0 row"
          >
            <div
              class="container-body d-flex flex-column justify-content-center align-items-center container"
            >
              <form
                class="payment-form"
              >
                <div
                  id="card-element"
                />
                <button
                  disabled=""
                  id="submit"
                >
                  <span
                    id="button-text"
                  >
                    Pay
                  </span>
                </button>
                
                <p
                  class="result-message"
                >
                  Payment succeeded, see the result in your
                  <a
                    href="https://dashboard.stripe.com/test/payments"
                  >
                     Stripe dashboard.
                  </a>
                   
                  Refresh the page to pay again.
                </p>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="p-0 jumbotron jumbotron-fluid"
    >
      <div
        class="container-fluid"
      >
        <div
          class="top-banner row"
        />
        <div
          class="reservation-wrapper px-0 row"
        >
          <div
            class="container-body d-flex flex-column justify-content-center align-items-center container"
          >
            <form
              class="payment-form"
            >
              <div
                id="card-element"
              />
              <button
                disabled=""
                id="submit"
              >
                <span
                  id="button-text"
                >
                  Pay
                </span>
              </button>
              
              <p
                class="result-message"
              >
                Payment succeeded, see the result in your
                <a
                  href="https://dashboard.stripe.com/test/payments"
                >
                   Stripe dashboard.
                </a>
                 
                Refresh the page to pay again.
              </p>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
